File: src/net/dinkla/raytracer/math/Point3DF.scala
package net.dinkla.raytracer.math

/**
  * Created by Dinkla on 10.07.2015.
  */
final class Point3DF(x2: Float, y2: Float, z2: Float) extends Point3D[Float](x2, y2, z2) {

   override val x: Float = x2
   override val y: Float = y2
   override val z: Float = z2

   override def length: Float = math.sqrt(sqrLength).toFloat

   override def unary_-(): Point3D[Float] = new Point3DF(-x, -y, -z)

   //override def plus(v: Vector3D[Float]): Point3D[Float] = new Point3DF(x + v.x, y + v.y, z + v.z)

  override def plus(p: Point3D[Float]): Vector3DF = new Vector3DF(x + p.x, y + p.y, z + p.z)
  override def +(v: Vector3D[Float]): Point3DF = plus(v)

   override def plus(f: Float): Point3DF = new Point3DF(x + f, y + f, z + f)

   override def toVector: Vector3DF = new Vector3DF(x, y, z)

  // temp
  override def plus(v: Vector3D[Float]): Point3DF = new Point3DF(x + v.x, y + v.y, z + v.z)

  override def minus(p: Point3D[Float]): Vector3DF = new Vector3DF(x - p.x, y - p.y, z - p.z)

  // Point minus Point gives a Vector
  override def minus(f: Float): Point3DF = new Point3DF(x - f, y - f, z - f)

  override def getX: Float = x

  override def getY: Float = y

  override def getZ: Float = z

}

File: src/net/dinkla/raytracer/math/Point3DF.java
//package net.dinkla.raytracer.math;
//
//public class Point3DF extends Element3D {
//
//    private float x;
//    private float y;
//    private float z;
//
//    static public final Point3DF ORIGIN = new Point3DF(0, 0, 0);
//    static public final Point3DF MAX = new Point3DF(Float.POSITIVE_INFINITY, Float.POSITIVE_INFINITY, Float.POSITIVE_INFINITY);
//    static public final Point3DF MIN = new Point3DF(Float.NEGATIVE_INFINITY, Float.NEGATIVE_INFINITY, Float.NEGATIVE_INFINITY);
//    static public final Point3DF DEFAULT_CAMERA = new Point3DF(0, 10, 10);
//
//    public Point3DF(final float x, final float y, final float z) {
//        super(x, y, z);
//    }
//
//    public Point3DF(final Element3D e) {
//        super(e);
//    }
//
//    public Vector3DF plus(final Point3DF v) {
//        return new Vector3DF(getX() + v.getX(), getY() + v.getY(), getZ() + v.getZ());
//    }
//
//    public Point3DF plus(final Vector3DF v) {
//        return new Point3DF(getX() + v.x, getY() + v.y, getZ() + v.z);
//    }
//
//    public Point3DF plus(final float f) {
//        return new Point3DF(getX() + f, getY() + f, getZ() + f);
//    }
//
//    public Vector3DF minus(final Point3DF v) {
//        return new Vector3DF(getX() - v.getX(), getY() - v.getY(), getZ() - v.getZ());
//    }
//
//    public Point3DF minus(final Vector3DF v) {
//        return new Point3DF(getX() - v.x, getY() - v.y, getZ() - v.z);
//    }
//
//    public Point3DF minus(final float f) {
//        return new Point3DF(getX() - f, getY() - f, getZ() - f);
//    }
//
//    public float getX() {
//        return x;
//    }
//
//    public void setX(float x) {
//        this.x = x;
//    }
//
//    public float getY() {
//        return y;
//    }
//
//    public void setY(float y) {
//        this.y = y;
//    }
//
//    public float getZ() {
//        return z;
//    }
//
//    public void setZ(float z) {
//        this.z = z;
//    }
//}
